src/taxpasta/domain/service/sample_merging_service.py

```diff
@@ -0,0 +1,87 @@
+# flake8: noqa
+# Copyright (c) 2022, Moritz E. Beber, Maxime Borry, Jianhong Ou, Sofia Stamouli.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Provide a sample merging service that summarizes two or more samples."""
+
+
+from typing import Iterable
+
+import pandas as pd
+import pandera as pa
+from pandera.typing import DataFrame
+
+from ..model import ObservationMatrix, Sample, StandardProfile, TidyObservationTable
+
+
+class SampleMergingService:
+    """Define a sample merging service that summarizes one or more samples."""
+
+    @classmethod
+    @pa.check_types(lazy=True)
+    def merge_wide(cls, samples: Iterable[Sample]) -> DataFrame[ObservationMatrix]:
+        """
+        Merge two or more sample profiles into a wide-format observation matrix.
+
+        Args:
+            samples: Two or more samples.
+
+        Returns:
+            A single table containing one row per taxon, one column for the taxonomy
+            identifier, and one column per sample with abundance counts.
+
+        """
+        # `set_index` creates a copy of the original profile which is convenient so that
+        # we do not modify existing profiles but, of course, doubles the memory used.
+        counts = [
+            sample.profile.set_index(
+                keys=StandardProfile.taxonomy_id, verify_integrity=True
+            ).rename(columns={StandardProfile.count: sample.name})
+            for sample in samples
+        ]
+        # Please note that `set_index` restores the underlying dtype of the categorical
+        # column `taxonomy_id`. Thus, when we `reset_index` the column is of dtype
+        # object but, due to schema coercion, this is automatically converted into a
+        # categorical dtype again when pandera checks the return type.
+        return (
+            counts[0].join(counts[1:], how="outer").fillna(0).astype(int).reset_index()
```

Comment:
Maybe we need to add a check for NaN, and warn is some are found instead of directly replacing them by 0.
I'm thinkink of the implications of 0s https://www.sciencedirect.com/science/article/pii/S2001037020303986

Reply:
Hmm, while I agree with the discussion in the paper, I don't know of any downstream analysis method that distinguishes consistently between zeros and NAs. Also, some tools (like kraken2) have the option to output all non-identified taxa with a zero count. So I technically agree but I am not convinced that up- or downstream tools would handle this correctly anyway.

Reply:
I agree with you, I also don't differentiate them neither. But a warning would be nice, so as to remind the user of the different kind of zeros